dev/AnimatedIcon/AnimatedIcon.cpp

```diff
@@ -59,17 +59,35 @@ void AnimatedIcon::OnApplyTemplate()
 
 void AnimatedIcon::OnLoaded(winrt::IInspectable const&, winrt::RoutedEventArgs const&)
 {
-    // AnimatedIcon might get added to a UI which has already set the State property on the parent.
+    // AnimatedIcon might get added to a UI which has already set the State property on an ancestor.
     // If this is the case and the animated icon being added doesn't have its own state property
-    // We copy the parent value when we load.
+    // We copy the ancestor value when we load. Additionally we attach to our ancestor's property
+    // changed event for AnimatedIcon.State to copy the value to AnimatedIcon.
     auto const property = winrt::AnimatedIcon::StateProperty();
-    auto const stateValue = GetValue(property);
-    if (unbox_value<winrt::hstring>(stateValue).empty())
+
+    auto const [ancestorWithState, stateValue] = [this, property]()
     {
-        if (auto const parent = winrt::VisualTreeHelper::GetParent(*this))
+        auto parent = winrt::VisualTreeHelper::GetParent(*this);
+        while (parent)
         {
-            SetValue(property, parent.GetValue(property));
+            auto const stateValue = parent.GetValue(property);
+            if (!unbox_value<winrt::hstring>(stateValue).empty())
+            {
+                return std::make_tuple(parent, stateValue);
+            }
+            parent = winrt::VisualTreeHelper::GetParent(parent);
         }
+        return std::make_tuple(static_cast<winrt::DependencyObject>(nullptr), winrt::box_value(winrt::hstring{}));
+    }();
+
+    if (unbox_value<winrt::hstring>(GetValue(property)).empty())
+    {
+        SetValue(property, stateValue);
+    }
+
+    if (ancestorWithState)
+    {
+        m_ancestorStatePropertyChangedRevoker = RegisterPropertyChanged(ancestorWithState, property, { this, &AnimatedIcon::OnAncestorAnimatedIconStatePropertyChanged });
```

Comment:
do you need to revoke first if this gets called multiple times?

Reply:
Revokers are like smart pointers -- they'll unregister from the old event when you reassign them, if that's what you're asking?

Reply:
Correct, you don't need to revoke before reassigning.